File: orchestrator/policy.py
"""The authoritative decision table (finalplan §4).

Pure function, no side effects: given the anomaly score and the incident severity,
return the containment action + whether a human must approve it. The agent's
`suggested_action` is only a hint — this is what actually decides.
"""

from typing import NamedTuple

from orchestrator.schemas import ActionType

HIGH_SEVERITIES = {"critical", "high"}


class Decision(NamedTuple):
    action: ActionType
    requires_human_approval: bool


def decide(anomaly_score: float, severity: str) -> Decision:
    if severity in HIGH_SEVERITIES and anomaly_score >= 0.9:
        return Decision(ActionType.isolate_host, True)
    if anomaly_score >= 0.7:
        return Decision(ActionType.block_ip, False)
    return Decision(ActionType.monitor, False)

File: frontend/src/data/policy.ts
/**
 * Client-side mirror of orchestrator/policy.py — the AUTHORITATIVE decision table (README §4).
 * The agent's `suggested_action` is only a hint; this maps (anomaly_score, severity) to the
 * final action + approval requirement. Keep it a readable table.
 *
 *   severity ∈ {critical, high} AND score ≥ 0.9  -> isolate_host   (human approval required)
 *   score 0.7 – 0.9                              -> block_ip       (auto)
 *   score < 0.7                                  -> monitor        (auto)
 *
 * NOTE: in live mode the backend emits the ContainmentAction directly; this local copy is
 * only used by MockDataService to decide actions. If policy.py changes, update this too.
 */
import type { AnomalyEvent, EnrichedIncident, ActionType } from "../types/contracts";

export interface PolicyDecision {
  action: ActionType;
  target: string;
  requires_human_approval: boolean;
}

export function decide(event: AnomalyEvent, incident: EnrichedIncident): PolicyDecision {
  const score = event.anomaly_score;
  const highSeverity = incident.severity === "critical" || incident.severity === "high";

  if (highSeverity && score >= 0.9) {
    // Highest blast radius: quarantine the compromised internal host, behind approval.
    return { action: "isolate_host", target: event.src_ip, requires_human_approval: true };
  }
  if (score >= 0.7) {
    // Block the external counterparty automatically.
    return { action: "block_ip", target: event.dst_ip, requires_human_approval: false };
  }
  return { action: "monitor", target: event.src_ip, requires_human_approval: false };
}

/** Actions the policy runs without a human (drives automation-coverage metric). */
export function isAutonomous(action: ActionType): boolean {
  return action === "monitor" || action === "block_ip";
}
